regproxy/controller.go

```diff
@@ -0,0 +1,227 @@
+	go func() {
+		err := c.stopOldDarwinProxies(ctx)
+		if err != nil {
+			c.cons.Warnf("Failed to stop old Darwin proxy support container: %s", err)
+		}
+	}()
+
+	containerPort, err := acquireFreePort(ctx)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to acquire free port")
+	}
+
+	runCfg := containerutil.ContainerRun{
+		NameOrID: containerName,
+		ImageRef: fmt.Sprintf("alpine/socat:%s", c.darwinProxyImageTag),
+		Ports: []containerutil.Port{
+			{
+				IP:            "127.0.0.1",
+				HostPort:      containerPort, // Bind to available port
+				ContainerPort: 80,
+				Protocol:      containerutil.ProtocolTCP,
+			},
+		},
+		ContainerArgs: []string{
+			"tcp-listen:80,fork,reuseaddr",
+			fmt.Sprintf("tcp:host.docker.internal:%d", registryPort),
+		},
+	}
+
+	err = c.containerFrontend.ContainerRun(ctx, runCfg)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to start support container")
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, c.darwinProxyWait)
+	defer cancel()
+
+	// Wait for the proxy chain to resolve to the BK registry. The /v2/ path
+	// will return a 200 when ready.
+	for {
+		res, err := http.Get(fmt.Sprintf("http://127.0.0.1:%d/v2/", containerPort))
+		if res != nil && res.Body != nil {
+			res.Body.Close()
+		}
+		if err == nil && res != nil && res.StatusCode == http.StatusOK {
+			break
+		}
+		select {
+		case <-ctx.Done():
+			return 0, ctx.Err()
+		case <-time.After(time.Second):
+			continue
+		}
+	}
+
+	return containerPort, nil
+}
+
+func (c *Controller) stopOldDarwinProxies(ctx context.Context) error {
+	containers, err := c.containerFrontend.ContainerList(ctx)
+	if err != nil {
+		return err
+	}
+	for _, container := range containers {
+		if strings.HasPrefix(container.Name, darwinContainerPrefix) &&
+			time.Now().Sub(container.Created) > darwinContainerMaxAge {
+			err = c.stopDarwinProxy(ctx, container.Name, false)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (c *Controller) stopDarwinProxy(ctx context.Context, containerName string, checkExists bool) error {
+	// Ignore parent context cancellations as to prevent orphaned containers.
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	if checkExists {
+		infos, err := c.containerFrontend.ContainerInfo(ctx, containerName)
+		if err != nil {
+			return err
+		}
+		if info, ok := infos[containerName]; !ok || info.Status == containerutil.StatusMissing {
+			return nil
+		}
+	}
+	err := c.containerFrontend.ContainerRemove(ctx, true, containerName)
+	if err != nil {
+		return errors.Wrap(err, "failed to stop support container")
+	}
+	return nil
+}
+
+func acquireFreePort(ctx context.Context) (int, error) {
+	addr := "127.0.0.1:0"
+
+	ln, err := (&net.ListenConfig{}).Listen(ctx, "tcp", addr)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to listen on open port")
+	}
+	defer ln.Close() // Immediately close the listener
+
+	return ln.Addr().(*net.TCPAddr).Port, nil
+}
```

Comment:
I noticed that as well. We can use a file based lock to prevent the race across `earthly` process instances, but that's a bit combersome. Although, It does seem exceedingly unlikely that someone would hit this and the fix would just be a retry. Thoughts?

Reply:
specifically binding to port `0` asks the OS to pick an unused port.

Reply:
Right, but 2 `earthly` builds could identify the same free port if started in parallel. A listens and closes, B listens and closes, both attempt to use port for container. Still, quite unlikely.

Reply:
> specifically binding to port `0` asks the OS to pick an unused port.

I was using `0` originaly for the container port, explicitly. But there's an issue with [this code](https://github.com/earthly/earthly/blob/main/util/containerutil/shell_shared.go#L179) on Mac. The empty string differs from `0` in Docker Destop.

Reply:
I missed the part that `ln.Close()` is being called.... I understand now.

Is there a way to have `docker run .....` dynamically pick the port, then we call `docker inspect` to figure out what it was assigned?

Reply:
Yes, but see my previous comment: https://github.com/earthly/earthly/pull/3420#discussion_r1372163696

I switched to this method after encountering a difficult to identify bug.